### ISSUE TITLE
Overeager type narrowing

### ISSUE THREAD

**Issue by kmillikin** [microsoft/pyright#4094]
I'm implementing type narrowing and thinking about intersection types for Python in general and I noticed this behavior in Pyright:

```python
Point = tuple[int, int]

class Shape: pass

def scale(x: float, o: int | Point | Shape):
  if isinstance(o, tuple):
    reveal_type(o)
```

Will show `tuple[int, int]` which might seem reasonable but what if there's

```python
class ColorShape(Shape, tuple[Shape, str]): pass
```

We should say that the intersection of `Point` and `tuple` (= `tuple[Any, ...]`) is `Point` (= `tuple[int, int]`). The intersection of `Shape` and `tuple` is not known to be empty.

The intersection of `int` and `tuple` is known to be empty for CPython because:

```
>>> class A(int, tuple): pass
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: multiple bases have instance lay-out conflict
```

**Comment by kmillikin** [comment]
To be more explicit, I think that the type of `o` there is really `tuple[int, int] | <subclass of Shape and tuple[Any, ...]>`.

**Comment by erictraut** [comment]
The current behavior is intended, not a bug. This is an example of practicality over purity. It's extremely rare that someone would use multiple inheritance to combine a `Shape` and `tuple`, but it's extremely common for users to want to narrow a union type using `isinstance` checks. Other languages like TypeScript (which supports unions) make similar assumptions in their narrowing logic.

Even if intersection types were formally added to the language, I'd be reluctant to change this behavior.

**Comment by kmillikin** [comment]
I wasn't specifically talking about tuples.

What I means was, if `x: A | B` and we narrow it with `isinstance(x, A)` then what we can conclude is that the narrowed uses of `x` will have type `A | <subclass of A and B>`.

I don't quite see how it's a usability problem.  I _think_ what users want to do after code like `isinstance(x, A)` is to invoke methods of `A` on it, and they can still do that.

**Comment by erictraut** [comment]
If you ask Python developers what they expect to see when narrowing the type `A | B` with an `isinstance(x, A)` check, I would be willing to bet that 99%+ would expect the answer to be `A`. It would be surprising and confusing to the vast majority if a tool like pyright or pylance would display the type `A | <subclass of A and B>`. That would defy expectations and would not be their intent the vast vast majority of the time. So while you are _technically_ correct here, this would not be good for usability.

**Comment by kmillikin** [comment]
That's probably right, but here I'm talking about how we narrow the type `A | B` with an `isinstance(x, C)` check.

I don't think 99% of Python programmers would necessarily say `A` :)

**Comment by kmillikin** [comment]
The only problem I was trying to solve was to report the issue.  I wasn't sure it was intentional.  As I said, I'm implementing narrowing and I noticed this.

The issue is based on real-world code that is very much like the first example I posted.  The programmer has forgotten (or did not know) that there could be subclasses of `tuple` and so they have assumed that by checking that an expression was an instance of `tuple` then they would know the length of the tuple and the types of the components.  But that code has a type error because it isn't prepared to handle other sorts of tuples.

I'm not sure what percentage of Python programmers would want to know about the type error.  And I'm kind of a hard-liner in the sense that I'm not worried about defying a programmer's expectation if that expectation is wrong.  That will help them learn!

I don't know TypeScript well enough to know how the type system works.  In my test, which I though was exactly your example above (with interfaces and a user-defined typeguard for `C`), I observed a narrowed type of `C`, not `A` using the playground at typescript.org.  Not sure if that's the same thing or the best way to test it.

My implementation is admittedly different because it will collapse `A | <subclass of A and B>` into just `A`.  (Whereas collapsing `A | <subclass of B and C>` into `A` in some cases seems to go against the other principle of not throwing away information from unions.)
